tangelo/toolboxes/circuits/tests/test_lcu.py

```diff
@@ -0,0 +1,187 @@
+        phase estimation.
+        """
+
+        # Generate qubit operator with state 9 having eigenvalue 0.25
+        qu_op = (QubitOperator("X0 X1", 0.125) + QubitOperator("Y1 Y2", 0.125) + QubitOperator("Z2 Z3", 0.125)
+                 + QubitOperator("", 0.125))
+
+        ham_mat = get_sparse_operator(qu_op).toarray()
+        _, wavefunction = np.linalg.eigh(ham_mat)
+
+        qu_op = qu_op
+        # Kronecker product 13 qubits in the zero state to eigenvector 9 to account for ancilla qubits
+        wave_9 = wavefunction[:, 9]
+        for i in range(13):
+            wave_9 = np.kron(wave_9, np.array([1, 0]))
+
+        qubit_list = [16, 15, 14]
+
+        pe_circuit = get_qft_circuit(qubit_list)
+        for i, qubit in enumerate(qubit_list):
+            pe_circuit += get_truncated_taylor_series(qu_op, 3, t=-(2*np.pi)*2**i, control=qubit)
+        pe_circuit += get_qft_circuit(qubit_list, inverse=True)
+
+        freqs, _ = sim_cirq.simulate(pe_circuit, initial_statevector=wave_9)
+
+        # Trace out all but final 3 indices
+        trace_freq = dict()
+        for key, value in freqs.items():
+            trace_freq[key[-3:]] = trace_freq.get(key[-3:], 0) + value
+
+        # State 9 has eigenvalue 0.25 so return should be 010 (0*1/2 + 1*1/4 + 0*1/8)
+        self.assertAlmostEqual(trace_freq["010"], 1.0, delta=0.0001)
+
+    def test_controlled_time_evolution_by_phase_estimation_for_get_lcu_circuit(self):
+        """ Verify that the controlled time-evolution is correct by calculating the eigenvalue of an eigenstate through
+        phase estimation.
+        """
+
+        # Generate qubit operator with state 9 having eigenvalue 0.25
+        qu_op = (QubitOperator("X0 X1", 0.125) + QubitOperator("Y1 Y2", 0.125) + QubitOperator("Z2 Z3", 0.125)
+                 + QubitOperator("", 0.125))
+
+        ham_mat = get_sparse_operator(qu_op).toarray()
+        _, wavefunction = np.linalg.eigh(ham_mat)
+
+        qu_op = qu_op
+
+        # break time into 6 parts so 1-norm is less than 2. i.e. can use Oblivious Amplitude Amplification
+        time = -2 * np.pi / 6
+
+        # Generate explicit qubit operator exponential
+        exp_qu_op = 1 + -1j*qu_op*time + (-1j*qu_op*time)**2/2 + (-1j*qu_op*time)**3/6
+
+        num_terms = len(exp_qu_op.terms)
+        n_extra_qubits = math.ceil(np.log2(num_terms))
+
+        # Kronecker product n_extra_qubits + 3 qubits in the zero state to eigenvector 9
+        wave_9 = wavefunction[:, 9]
+        for i in range(n_extra_qubits+3):
+            wave_9 = np.kron(wave_9, np.array([1, 0]))
+
+        # Phase estimation register in reversed order
+        qubit_list = list(reversed(range(4 + n_extra_qubits, 4 + n_extra_qubits + 3)))
+
+        # Build phase estimation circuit and simulate
+        pe_circuit = get_qft_circuit(qubit_list)
+        for i, qubit in enumerate(qubit_list):
+            pe_circuit += get_lcu_circuit(exp_qu_op, control=qubit) * 2 ** i * 6
```

Comment:
I suggest `6 * 2**i * get_lcu_circuit(...)`, which is typically how I'd write formulas on paper.
This assumes btw the `*` operator works both ways, which I think it does?